File: tools/covenant/executor.py
#!/usr/bin/env python3
"""Executor policy engine for Covenant orchestration v2."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

HARD_ESCALATE_FAILURES = {"auth_failure", "permission_denied", "requirements_ambiguous"}
TRANSIENT_FAILURES = {"transient_tool_failure", "network_timeout", "timeout"}

ROUTE_SUGGESTION = {
    "agent.researcher.v1": "agent.oracle.v1",
    "agent.oracle.v1": "agent.librarian.v1",
    "agent.monitor.v1": "agent.huragok.v1",
    "agent.huragok.v1": "agent.monitor.v1",
    "agent.librarian.v1": "agent.huragok.v1",
}


def _step_index(plan: dict[str, Any]) -> dict[str, dict[str, Any]]:
    return {s.get("step_id"): s for s in plan.get("steps", []) if isinstance(s, dict) and isinstance(s.get("step_id"), str)}


def _parallel_groups(plan: dict[str, Any]) -> dict[str, set[str]]:
    groups: dict[str, set[str]] = {}
    for step in plan.get("steps", []):
        if not isinstance(step, dict):
            continue
        sid = step.get("step_id")
        grp = step.get("parallel_group")
        if isinstance(sid, str) and isinstance(grp, str) and grp.strip():
            groups.setdefault(grp.strip(), set()).add(sid)
    return groups


def group_for_step(plan: dict[str, Any], step_id: str) -> str | None:
    step = _step_index(plan).get(step_id)
    grp = step.get("parallel_group") if step else None
    return grp.strip() if isinstance(grp, str) and grp.strip() else None


def group_is_complete(plan: dict[str, Any], parallel_group: str, completed_steps: set[str]) -> bool:
    groups = _parallel_groups(plan)
    members = groups.get(parallel_group, set())
    return bool(members) and members.issubset(completed_steps)


def _expanded_dependencies(plan: dict[str, Any], step: dict[str, Any]) -> set[str]:
    """Expand dependencies so any dependency on one member of a parallel group gates on the full group."""
    deps = set(step.get("depends_on", []))
    idx = _step_index(plan)
    groups = _parallel_groups(plan)
    expanded = set(deps)

    for dep in deps:
        dep_step = idx.get(dep)
        grp = dep_step.get("parallel_group") if dep_step else None
        if isinstance(grp, str) and grp.strip() and grp.strip() in groups:
            expanded.update(groups[grp.strip()])

    return expanded


def next_ready_steps(plan: dict[str, Any], completed_steps: set[str]) -> list[dict[str, Any]]:
    """Return dispatch-ready steps.

    - Standard chain: returns one step.
    - Parallel group: when first ready step is in a group, return all ready members in that group.
    """
    steps = [s for s in plan.get("steps", []) if isinstance(s, dict)]
    idx = _step_index(plan)

    for step in steps:
        sid = step.get("step_id")
        if not isinstance(sid, str) or sid in completed_steps:
            continue

        deps = _expanded_dependencies(plan, step)
        if not deps.issubset(completed_steps):
            continue

        grp = step.get("parallel_group")
        if isinstance(grp, str) and grp.strip():
            group_name = grp.strip()
            ready: list[dict[str, Any]] = []
            for member_id, member in idx.items():
                member_group = member.get("parallel_group")
                if member_id in completed_steps:
                    continue
                if not (isinstance(member_group, str) and member_group.strip() == group_name):
                    continue
                member_deps = _expanded_dependencies(plan, member)
                if member_deps.issubset(completed_steps):
                    ready.append(member)
            if ready:
                return ready

        return [step]

    return []


def next_ready_step(plan: dict[str, Any], completed_steps: set[str]) -> dict[str, Any] | None:
    ready = next_ready_steps(plan, completed_steps)
    return ready[0] if ready else None


def decide_retry(agent_identity_id: str, failure_type: str, attempt: int, max_retries: int) -> dict[str, Any]:
    failure_type = failure_type.strip().lower()

    if failure_type in HARD_ESCALATE_FAILURES:
        return {
            "action": "escalate_immediately",
            "route_to": None,
            "reason": "Hard-blocking failure class; retries are unsafe by policy.",
        }

    if failure_type in TRANSIENT_FAILURES and attempt <= max_retries:
        return {
            "action": "retry_same_agent",
            "route_to": agent_identity_id,
            "reason": "Transient/timeout failure within retry budget.",
        }

    return {
        "action": "escalate_with_route_suggestion",
        "route_to": ROUTE_SUGGESTION.get(agent_identity_id),
        "reason": "Failure exceeded retry budget or non-transient class.",
    }


def build_execution_state(
    plan: dict[str, Any],
    critique: dict[str, Any],
    completed_steps: set[str] | None = None,
    failure_event: dict[str, Any] | None = None,
) -> dict[str, Any]:
    completed_steps = completed_steps or set()

    if not critique.get("approved"):
        return {
            "state": "blocked",
            "current_step_id": None,
            "dispatch_step_ids": [],
            "next_action": "replan_or_human_review",
            "retry_decision": {
                "action": "none",
                "route_to": None,
                "reason": "Critic rejected plan; execution halted before dispatch.",
            },
        }

    if failure_event:
        retry = decide_retry(
            failure_event["agent_identity_id"],
            failure_event["failure_type"],
            int(failure_event["attempt"]),
            int(failure_event["max_retries"]),
        )
        return {
            "state": "blocked" if retry["action"].startswith("escalate") else "running",
            "current_step_id": failure_event.get("step_id"),
            "dispatch_step_ids": [failure_event.get("step_id")] if failure_event.get("step_id") else [],
            "next_action": retry["action"],
            "retry_decision": retry,
        }

    ready_steps = next_ready_steps(plan, completed_steps)
    if not ready_steps:
        return {
            "state": "completed",
            "current_step_id": None,
            "dispatch_step_ids": [],
            "next_action": "final_quality_gate",
            "retry_decision": {"action": "none", "route_to": None, "reason": "All steps complete."},
        }

    dispatch_ids = [s["step_id"] for s in ready_steps if isinstance(s.get("step_id"), str)]
    return {
        "state": "running",
        "current_step_id": dispatch_ids[0] if dispatch_ids else None,
        "dispatch_step_ids": dispatch_ids,
        "next_action": "dispatch_step",
        "retry_decision": {"action": "none", "route_to": None, "reason": "No failure event."},
    }


def main() -> None:
    import argparse

    parser = argparse.ArgumentParser(description="Build executor state for Covenant plan")
    parser.add_argument("plan", help="Path to plan JSON")
    parser.add_argument("critique", help="Path to critique JSON")
    parser.add_argument("--completed", help="JSON array file of completed step ids")
    parser.add_argument("--failure", help="Failure event JSON path")
    args = parser.parse_args()

    plan = json.loads(Path(args.plan).expanduser().resolve().read_text())
    critique = json.loads(Path(args.critique).expanduser().resolve().read_text())
    completed: set[str] = set()
    if args.completed:
        completed = set(json.loads(Path(args.completed).expanduser().resolve().read_text()))
    failure = json.loads(Path(args.failure).expanduser().resolve().read_text()) if args.failure else None

    print(json.dumps(build_execution_state(plan, critique, completed, failure), indent=2))


if __name__ == "__main__":
    main()

File: tools/covenant/executor.ts
#!/usr/bin/env npx tsx

/** Executor policy engine for Covenant orchestration v2. */

import fs from "fs";
import path from "path";

const HARD_ESCALATE_FAILURES = new Set(["auth_failure", "permission_denied", "requirements_ambiguous"]);
const TRANSIENT_FAILURES = new Set(["transient_tool_failure", "network_timeout", "timeout"]);

const ROUTE_SUGGESTION: Record<string, string | undefined> = {
  "agent.researcher.v1": "agent.oracle.v1",
  "agent.oracle.v1": "agent.librarian.v1",
  "agent.monitor.v1": "agent.huragok.v1",
  "agent.huragok.v1": "agent.monitor.v1",
  "agent.librarian.v1": "agent.huragok.v1",
};

type Json = Record<string, unknown>;

function stepIndex(plan: Json): Record<string, Json> {
  const steps = plan.steps;
  if (!Array.isArray(steps)) return {};
  const out: Record<string, Json> = {};
  for (const step of steps) {
    if (step && typeof step === "object") {
      const stepObj = step as Json;
      if (typeof stepObj.step_id === "string") {
        out[stepObj.step_id] = stepObj;
      }
    }
  }
  return out;
}

function parallelGroups(plan: Json): Record<string, Set<string>> {
  const groups: Record<string, Set<string>> = {};
  const steps = plan.steps;
  if (!Array.isArray(steps)) return groups;

  for (const step of steps) {
    if (!step || typeof step !== "object") continue;
    const stepObj = step as Json;
    const sid = stepObj.step_id;
    const grp = stepObj.parallel_group;
    if (typeof sid === "string" && typeof grp === "string" && grp.trim()) {
      const key = grp.trim();
      if (!groups[key]) groups[key] = new Set();
      groups[key].add(sid);
    }
  }
  return groups;
}

export function groupForStep(plan: Json, stepId: string): string | null {
  const step = stepIndex(plan)[stepId];
  const grp = step ? step.parallel_group : null;
  if (typeof grp === "string" && grp.trim()) return grp.trim();
  return null;
}

export function groupIsComplete(plan: Json, parallelGroup: string, completedSteps: Set<string>): boolean {
  const groups = parallelGroups(plan);
  const members = groups[parallelGroup] ?? new Set<string>();
  return members.size > 0 && Array.from(members).every((m) => completedSteps.has(m));
}

function expandedDependencies(plan: Json, step: Json): Set<string> {
  const depsRaw = step.depends_on;
  const deps = Array.isArray(depsRaw) ? depsRaw.filter((d) => typeof d === "string") : [];
  const idx = stepIndex(plan);
  const groups = parallelGroups(plan);
  const expanded = new Set<string>(deps as string[]);

  for (const dep of deps as string[]) {
    const depStep = idx[dep];
    const grp = depStep ? depStep.parallel_group : null;
    if (typeof grp === "string" && grp.trim() && groups[grp.trim()]) {
      for (const member of groups[grp.trim()]) {
        expanded.add(member);
      }
    }
  }
  return expanded;
}

export function nextReadySteps(plan: Json, completedSteps: Set<string>): Json[] {
  const stepsRaw = plan.steps;
  const steps = Array.isArray(stepsRaw) ? stepsRaw.filter((s) => s && typeof s === "object") : [];
  const idx = stepIndex(plan);

  for (const step of steps) {
    const stepObj = step as Json;
    const sid = stepObj.step_id;
    if (typeof sid !== "string" || completedSteps.has(sid)) continue;

    const deps = expandedDependencies(plan, stepObj);
    let ready = true;
    for (const dep of deps) {
      if (!completedSteps.has(dep)) {
        ready = false;
        break;
      }
    }
    if (!ready) continue;

    const grp = stepObj.parallel_group;
    if (typeof grp === "string" && grp.trim()) {
      const groupName = grp.trim();
      const readyMembers: Json[] = [];
      for (const [memberId, member] of Object.entries(idx)) {
        const memberGroup = member.parallel_group;
        if (completedSteps.has(memberId)) continue;
        if (!(typeof memberGroup === "string" && memberGroup.trim() === groupName)) continue;
        const memberDeps = expandedDependencies(plan, member);
        let memberReady = true;
        for (const dep of memberDeps) {
          if (!completedSteps.has(dep)) {
            memberReady = false;
            break;
          }
        }
        if (memberReady) readyMembers.push(member);
      }
      if (readyMembers.length) return readyMembers;
    }

    return [stepObj];
  }

  return [];
}

export function nextReadyStep(plan: Json, completedSteps: Set<string>): Json | null {
  const ready = nextReadySteps(plan, completedSteps);
  return ready.length ? ready[0] : null;
}

export function decideRetry(agentIdentityId: string, failureType: string, attempt: number, maxRetries: number): Json {
  const failure = failureType.trim().toLowerCase();

  if (HARD_ESCALATE_FAILURES.has(failure)) {
    return {
      action: "escalate_immediately",
      route_to: null,
      reason: "Hard-blocking failure class; retries are unsafe by policy.",
    };
  }

  if (TRANSIENT_FAILURES.has(failure) && attempt <= maxRetries) {
    return {
      action: "retry_same_agent",
      route_to: agentIdentityId,
      reason: "Transient/timeout failure within retry budget.",
    };
  }

  return {
    action: "escalate_with_route_suggestion",
    route_to: ROUTE_SUGGESTION[agentIdentityId],
    reason: "Failure exceeded retry budget or non-transient class.",
  };
}

export function buildExecutionState(
  plan: Json,
  critique: Json,
  completedSteps: Set<string> | null = null,
  failureEvent: Json | null = null
): Json {
  const completed = completedSteps ?? new Set<string>();

  if (!critique.approved) {
    return {
      state: "blocked",
      current_step_id: null,
      dispatch_step_ids: [],
      next_action: "replan_or_human_review",
      retry_decision: {
        action: "none",
        route_to: null,
        reason: "Arbiter rejected plan; execution halted before dispatch.",
      },
    };
  }

  if (failureEvent) {
    const retry = decideRetry(
      String(failureEvent.agent_identity_id ?? ""),
      String(failureEvent.failure_type ?? ""),
      Number(failureEvent.attempt ?? 0),
      Number(failureEvent.max_retries ?? 0)
    );
    return {
      state: String(retry.action).startsWith("escalate") ? "blocked" : "running",
      current_step_id: failureEvent.step_id ?? null,
      dispatch_step_ids: failureEvent.step_id ? [failureEvent.step_id] : [],
      next_action: retry.action,
      retry_decision: retry,
    };
  }

  const readySteps = nextReadySteps(plan, completed);
  if (!readySteps.length) {
    return {
      state: "completed",
      current_step_id: null,
      dispatch_step_ids: [],
      next_action: "final_quality_gate",
      retry_decision: { action: "none", route_to: null, reason: "All steps complete." },
    };
  }

  const dispatchIds = readySteps
    .map((s) => s.step_id)
    .filter((id) => typeof id === "string") as string[];

  return {
    state: "running",
    current_step_id: dispatchIds[0] ?? null,
    dispatch_step_ids: dispatchIds,
    next_action: "dispatch_step",
    retry_decision: { action: "none", route_to: null, reason: "No failure event." },
  };
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  if (args.length < 2) {
    console.error("Usage: executor.ts <plan.json> <critique.json> [--completed <completed.json>] [--failure <failure.json>]");
    process.exit(2);
  }

  const plan = JSON.parse(fs.readFileSync(path.resolve(args[0]), "utf8"));
  const critique = JSON.parse(fs.readFileSync(path.resolve(args[1]), "utf8"));

  let completed: Set<string> = new Set();
  const completedIndex = args.indexOf("--completed");
  if (completedIndex !== -1 && args[completedIndex + 1]) {
    const raw = JSON.parse(fs.readFileSync(path.resolve(args[completedIndex + 1]), "utf8"));
    if (Array.isArray(raw)) {
      completed = new Set(raw.map((x) => String(x)));
    }
  }

  let failure: Json | null = null;
  const failureIndex = args.indexOf("--failure");
  if (failureIndex !== -1 && args[failureIndex + 1]) {
    failure = JSON.parse(fs.readFileSync(path.resolve(args[failureIndex + 1]), "utf8"));
  }

  const result = buildExecutionState(plan, critique, completed, failure);
  console.log(JSON.stringify(result, null, 2));
}

main();
